extra_data/compression.py

```diff
@@ -0,0 +1,122 @@
+import threading
+from copy import copy
+from multiprocessing.pool import ThreadPool
+
+import h5py
+import numpy as np
+from zlib_into import decompress_into
+
+
+def filter_ids(dset: h5py.Dataset):
+    dcpl = dset.id.get_create_plist()
+    return [dcpl.get_filter(i)[0] for i in range(dcpl.get_nfilters())]
+
+
+class DeflateDecompressor:
```

Comment:
It's a bit weird to me that we have a whole other class for deflate by itself, even though it explicitly requires a dataset with both shuffle and deflate filters to be constructed. Couldn't it be handled in `ShuffleDeflateDecompressor.apply_filters()` with this condition?
```python
if filter_mask & 1 and self.itemsize > 1:
```

(assuming the itemsize is stored too)

Reply:
It can be constructed from a dataset with only the deflate filter (`if filters == [h5py.h5z.FILTER_DEFLATE]:`), but I wanted to support the compressed gain data, where the shuffle filter is present but doesn't do anything, and I figured we may as well avoid the extra copy from the decompression buffer to the output array. That could be all in one class, but avoiding the extra copy would mean more code paths.

Also, I'm vaguely designing something that could be extended in the future if we start using different filters. I know that can be a risky thing to do, and I've tried not to go too far down the road of abstractions, but this seemed like a pretty easy and low cost one.

Reply:
Mmmm ok. I'm not fully convinced but I don't feel very strongly about it.